ui/selectors/selectors.js

```diff
@@ -477,43 +477,96 @@ export function getSelectedAccountNativeTokenCachedBalanceByChainId(state) {
 }
 
 /**
- * Based on the current account address, query for all tokens across all chain networks on that account.
- * This will eventually be exposed in a new piece of state called `tokenBalances`, which will including the new polling mechanism to stay up to date
+ * Based on the current account address, query for all tokens across all chain networks on that account,
+ * including the native tokens, without hardcoding any native token information.
  *
  * @param {object} state - Redux state
- * @returns {object} An array of tokens with balances for the given account. Data relationship will be chainId => balance
+ * @returns {object} An object mapping chain IDs to arrays of tokens (including native tokens) with balances.
  */
 export function getSelectedAccountTokensAcrossChains(state) {
   const { allTokens } = state.metamask;
-  const { address: selectedAddress } = getSelectedInternalAccount(state);
+  const selectedAddress = getSelectedInternalAccount(state).address;
 
-  // Initialize an empty object to hold tokens by chainId for the selected account
   const tokensByChain = {};
 
-  // Loop through each chain (0x1, 0x89, etc.)
-  for (const chainId in allTokens) {
-    if (allTokens[chainId]) {
-      const chainData = allTokens[chainId];
+  const nativeTokenBalancesByChainId =
+    getSelectedAccountNativeTokenCachedBalanceByChainId(state);
 
-      // Check if the selected account exists within this chain
-      if (chainData[selectedAddress]) {
-        // Ensure the chainId key exists in tokensByChain
-        if (!tokensByChain[chainId]) {
-          tokensByChain[chainId] = [];
-        }
+  const chainIds = new Set([
+    ...Object.keys(allTokens || {}),
+    ...Object.keys(nativeTokenBalancesByChainId || {}),
+  ]);
 
-        // Add each token to the array under its chainId
-        chainData[selectedAddress].forEach((token) => {
-          const tokenWithChain = { ...token, chainId };
-          tokensByChain[chainId].push(tokenWithChain);
-        });
-      }
+  chainIds.forEach((chainId) => {
+    if (!tokensByChain[chainId]) {
+      tokensByChain[chainId] = [];
     }
-  }
+
+    if (allTokens[chainId] && allTokens[chainId][selectedAddress]) {
```

Comment:
This can be:

```
if (allTokens[chainId]?.[selectedAddress]) {
```

Reply:
https://github.com/MetaMask/metamask-extension/pull/28193/commits/225108dd4ea13b9e7e5a2dc9970aed55db551d4a